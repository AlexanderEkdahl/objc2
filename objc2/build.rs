fn main() {
    // The script doesn't depend on our code
    println!("cargo:rerun-if-changed=build.rs");

    #[cfg(feature = "exception")]
    {
        use std::env;

        if std::env::var("DOCS_RS").is_ok() {
            // docs.rs doesn't have clang, so skip building this. The
            // documentation will still work since it doesn't need to link.
            return;
        }
        println!("cargo:rerun-if-changed=extern/exception.m");

        let mut builder = cc::Build::new();
        builder.file("extern/exception.m");

        for flag in env::var("DEP_OBJC_0_2_CC_ARGS").unwrap().split(' ') {
            builder.flag(flag);
        }

        builder.compile("librust_objc_try_catch_exception.a");
    }
}
