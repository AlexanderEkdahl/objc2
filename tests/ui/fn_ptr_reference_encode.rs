//! Test that `Encode` is not implemented for function pointers that are
//! higher-ranked over lifetimes.
//!
//! Ideally, they should be, but they can't be right now.
//!
//! (Also test that we can use `_` to work around this).
use objc2::Encode;

extern "C" fn my_fn(_x: &i32) {}

fn e<T: Encode>(_x: T) {}

fn main() {
    // Works
    e(my_fn as extern "C" fn(_));
    // Can't be written:
    // let encoding = <extern "C" fn(_) as Encode>::ENCODING;

    // Fails
    e(my_fn as extern "C" fn(&i32));
    // Also fails, properly tested in `fn_ptr_reference_encode2`
    let encoding = <extern "C" fn(&i32) as Encode>::ENCODING;
}
