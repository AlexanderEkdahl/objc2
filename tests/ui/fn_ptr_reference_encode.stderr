error: implementation of `Encode` is not general enough
  --> ui/fn_ptr_reference_encode.rs:13:5
   |
13 |     impls_encode(my_fn as extern "C" fn(_));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Encode` is not general enough
   |
   = note: `Encode` would have to be implemented for the type `for<'r> extern "C" fn(&'r i32)`
   = note: ...but `Encode` is actually implemented for the type `extern "C" fn(&'0 i32)`, for some specific lifetime `'0`

error: implementation of `Encode` is not general enough
  --> ui/fn_ptr_reference_encode.rs:18:5
   |
18 |     impls_encode(my_fn as extern "C" fn(&i32));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Encode` is not general enough
   |
   = note: `Encode` would have to be implemented for the type `for<'r> extern "C" fn(&'r i32)`
   = note: ...but `Encode` is actually implemented for the type `extern "C" fn(&'0 i32)`, for some specific lifetime `'0`

error: implementation of `Encode` is not general enough
  --> ui/fn_ptr_reference_encode.rs:19:21
   |
19 |     let _encoding = <extern "C" fn(&i32) as Encode>::ENCODING;
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Encode` is not general enough
   |
   = note: `Encode` would have to be implemented for the type `for<'r> extern "C" fn(&'r i32)`
   = note: ...but `Encode` is actually implemented for the type `extern "C" fn(&'0 i32)`, for some specific lifetime `'0`
